sql/core/src/main/scala/org/apache/spark/sql/streaming/ContinuousQueryListener.scala

```diff
@@ -69,27 +72,27 @@ abstract class ContinuousQueryListener {
 @Experimental
 object ContinuousQueryListener {
 
-  /**
-   * Base type of [[ContinuousQueryListener]] events.
-   * @since 2.0.0
-   */
+  /** Base type of [[ContinuousQueryListener]] events */
+  @JsonTypeInfo(
+    use = JsonTypeInfo.Id.CLASS,
+    include = JsonTypeInfo.As.PROPERTY,
+    property = "@class")
   trait Event
 
-  /**
-   * Event representing the start of a query.
-   * @since 2.0.0
-   */
-  class QueryStarted private[sql](val query: ContinuousQuery) extends Event
+  /** Event representing the start of a query */
+  class QueryStarted private[sql](val queryInfo: ContinuousQueryInfo) extends Event
 
-  /**
-   * Event representing any progress updates in a query.
-   * @since 2.0.0
-   */
-  class QueryProgress private[sql](val query: ContinuousQuery) extends Event
+  /** Event representing any progress updates in a query */
+  class QueryProgress private[sql](val queryInfo: ContinuousQueryInfo) extends Event
 
   /**
-   * Event representing that termination of a query.
-   * @since 2.0.0
+   * Event representing that termination of a query
+   *
+   * @param queryInfo
+   * @param exception The exception information of the [[ContinuousQuery]] if any. Otherwise, it
+   *                  will be `None`.
    */
-  class QueryTerminated private[sql](val query: ContinuousQuery) extends Event
+  class QueryTerminated private[sql](
+      val queryInfo: ContinuousQueryInfo,
+      val exception: Option[String]) extends Event
```

Comment:
I don't think a string is enough here.  I think you need the stack trace also for it to be useful.